### PR TITLE
Code cleanup, refactoring and organizing updates

### LINKED ISSUE #116: Enemy move revealed immediately as first rival fight begins
Version: v.0.5.2
Game: FRLG

Two different streamers have run into this issue where very rarely they see move information for a Pokemon they shouldn't know about. This appears to happen exclusively in lab fights, even before the Pokemon are sent out. I wonder if this has something to do with reading in game data and a desync between tracking it.  It's possible this bug affects other pokemon but is seen most commonly in lab fights.

image
Source: https://www.twitch.tv/videos/1536104846

Unsure if this affects games other than FRLG.

### PR DESCRIPTION
**Pull request by UTDZac**
Other than a huge refactoring of code for better organization (moved most globals to their respective file objects), this update includes some number of big fixes as well, notably:

- Battle data delay check to hopefully prevent information leak
   - fixes #116 _(i hope)_
- Your Pokemon's Trace ability autotracks abilities for you.
- The move summary for "Bounce" updated to mention paralysis chance.
- Minor fix for when the user deletes their `Settings.ini` file completely, the Tracker will now rebuild it without error.

### CODE REVIEW

File: ironmon_tracker/MoveData.lua [hunk 1]
```diff
@@ -1,3756 +1,3768 @@
...
 		name = "Bind",
-		type = PokemonTypes.NORMAL,
+		type = PokemonData.Types.NORMAL,
 		power = "15",
 		pp = "20",
 		accuracy = "75",
-		category = MoveCategories.PHYSICAL,
+		category = MoveData.Categories.PHYSICAL,
 		iscontact = true,
 		summary = "Deals damage and inflicts 1/16 of the target's max HP as damage for 2-5 turns.",
```
Comment (Fellshadow) [inline ironmon_tracker/MoveData.lua:313]: Could maybe also mention that [binding moves](https://bulbapedia.bulbagarden.net/wiki/Category:Binding_moves) such as bind also trap the target, preventing running away / switching out while the move is in effect

File: ironmon_tracker/MoveData.lua [hunk 2]
```diff
@@ -1,3756 +1,3768 @@
...
-		type = PokemonTypes.GROUND,
-		power = NOPOWER,
+		type = PokemonData.Types.GROUND,
+		power = Constants.NO_POWER,
 		pp = "20",
-		accuracy = PLACEHOLDER,
-		category = MoveCategories.STATUS,
+		accuracy = Constants.BLANKLINE,
+		category = MoveData.Categories.STATUS,
 		summary = "Creates a hazard for the enemy team (can stack 3 times). Switching out deals 1/8, 1/6, or 1/4 max HP in damage to Pokemon without Flying or Levitate.",
```
Comment (Fellshadow) [inline ironmon_tracker/MoveData.lua:2078]: I feel like the wording here could be mis-interpreted as "the pokemon that is switching out takes damage" rather than "pokemon take damage when switched in"

File: ironmon_tracker/MoveData.lua [hunk 3]
```diff
@@ -1,3756 +1,3768 @@
...
-		power = NOPOWER,
+		type = PokemonData.Types.FIGHTING,
+		power = Constants.NO_POWER,
 		pp = "5",
-		accuracy = PLACEHOLDER,
-		category = MoveCategories.STATUS,
+		accuracy = Constants.BLANKLINE,
+		category = MoveData.Categories.STATUS,
 		priority = "+ 3",
 		summary = "Protects user from all effects of moves for the turn, including damage. Using this move consectuively halves its success rate each time.",
```
Comment (Fellshadow) [inline ironmon_tracker/MoveData.lua:2139]: typo here: "consectuively" -> "consecutively"

File: ironmon_tracker/MoveData.lua [hunk 4]
```diff
@@ -1,3756 +1,3768 @@
...
 		name = "Rollout",
-		type = PokemonTypes.ROCK,
+		type = PokemonData.Types.ROCK,
 		power = "30",
 		pp = "20",
 		accuracy = "90",
-		category = MoveCategories.PHYSICAL,
+		category = MoveData.Categories.PHYSICAL,
 		iscontact = true,
 		summary = "Deals damage over 5 turns, doubling in power each consectutive hit. Base power is doubled if the user had previously used Defense Curl.",
```
Comment (Fellshadow) [inline ironmon_tracker/MoveData.lua:2222]: typo here: "consectutive" -> "consecutive" (also present on ice ball's summary)

File: ironmon_tracker/MoveData.lua [hunk 5]
```diff
@@ -1,3756 +1,3768 @@
...
-		type = PokemonTypes.NORMAL,
-		power = NOPOWER,
+		type = PokemonData.Types.NORMAL,
+		power = Constants.NO_POWER,
 		pp = "10",
-		accuracy = PLACEHOLDER,
-		category = MoveCategories.STATUS,
+		accuracy = Constants.BLANKLINE,
+		category = MoveData.Categories.STATUS,
 		summary = "Restores up to 50% of the user's max HP. Can be used outside of battle.",
```
Comment (Fellshadow) [inline ironmon_tracker/MoveData.lua:2253]: "Can be used outside of battle." could be misleading, as the out of battle effect is transferring user's hp to another pokemon not healing the user

File: ironmon_tracker/MoveData.lua [hunk 6]
```diff
@@ -1,3756 +1,3768 @@
...
-		type = PokemonTypes.ELECTRIC,
-		power = NOPOWER,
+		type = PokemonData.Types.ELECTRIC,
+		power = Constants.NO_POWER,
 		pp = "20",
-		accuracy = PLACEHOLDER,
-		category = MoveCategories.STATUS,
+		accuracy = Constants.BLANKLINE,
+		category = MoveData.Categories.STATUS,
 		summary = "User charges up, doubling the damage dealt by the next move it uses but only for Electric moves.",
```
Comment (Fellshadow) [inline ironmon_tracker/MoveData.lua:2879]: could maybe reword into "doubling the damage dealt by the next electric-type move"

File: ironmon_tracker/MoveData.lua [hunk 7]
```diff
@@ -1,3756 +1,3768 @@
...
 		name = "Metal Sound",
-		type = PokemonTypes.STEEL,
-		power = NOPOWER,
+		type = PokemonData.Types.STEEL,
+		power = Constants.NO_POWER,
 		pp = "40",
 		accuracy = "85",
-		category = MoveCategories.STATUS,
+		category = MoveData.Categories.STATUS,
 		summary = "Deals lowers the target's Special Defense stat by two stages. Fails against Pokemon with Soundproof.",
```
Comment (Fellshadow) [inline ironmon_tracker/MoveData.lua:3408]: typo here: "Deals lowers"

**Review by Fellshadow — CHANGES_REQUESTED** [review]
Looking great, there's just some tiny things here and there which i've pointed out (nothing major or breaking, just little things and a lot of it is typos in move summaries)

File: ironmon_tracker/Buttons.lua [hunk 8]
```diff
@@ -248,23 +255,28 @@ function Buttons.openAbilityNoteWindow()
...
 	local abilityOneDropdown = forms.dropdown(abilityForm, {["Init"]="Loading Ability1"}, 95, 30, 145, 30)
-	forms.setdropdownitems(abilityOneDropdown, MiscData.ability, true) -- true = alphabetize list
+	forms.setdropdownitems(abilityOneDropdown, abilityList, true) -- true = alphabetize list
 	local abilityTwoDropdown = forms.dropdown(abilityForm, {["Init"]="Loading Ability2"}, 95, 60, 145, 30)
-	forms.setdropdownitems(abilityTwoDropdown, MiscData.ability, true) -- true = alphabetize list
+	forms.setdropdownitems(abilityTwoDropdown, abilityList, true) -- true = alphabetize list
 
 	if trackedAbilities[1].id ~= 0 then
-		forms.settext(abilityOneDropdown, MiscData.ability[trackedAbilities[1].id + 1])
+		forms.settext(abilityOneDropdown, MiscData.Abilities[trackedAbilities[1].id])
```
[the remark this answers is not in the document]
Comment (UTDZac) [reply ironmon_tracker/Buttons.lua:276]: The old miscdata abilities had an extra "ability", which was just a blank placeholder. I removed all of these from pokemon/moves/abilities. Given this, I wanted this list to include a blank ability so the user can clear out an ability. This was added in manually on line 259. As a result, the `abilityList` object should only be used for displaying the ability names and retrieving an ability name. Anything involving the actual id needs to use the reference `MiscData.Abilities` list so the index will be correct.

File: ironmon_tracker/Buttons.lua [hunk 9]
```diff
@@ -73,19 +70,29 @@ AbilityTrackingButton = {
...
+	text = "",
+	textcolor = "Default text",
+	box = { Constants.SCREEN.WIDTH + 6, 141, 16, 16 },
+	onclick = function() 
+		Buttons.openNotePadWindow()
+	end
+}
+
 Buttons = {
 	{ -- HP button
```
[the remark this answers is not in the document]
Comment (UTDZac) [reply ironmon_tracker/Buttons.lua:90]: Big agree here. Part of this refactor process was to get those buttons cleaned up. I ended up change virtually everything else and exhausting myself in the process. I do still want to update these, but wanted to get some priority bug fixes in first.

File: ironmon_tracker/Buttons.lua (hunk 8 above)
[the remark this answers is not in the document]
Comment (Fellshadow) [reply ironmon_tracker/Buttons.lua:276]: yeah that makes sense 👍

**Review by Fellshadow — APPROVED** [review]
